llama-index-core/llama_index/core/workflow/workflow.py

```
@@ -78,6 +78,11 @@ def add_step(cls, func: Callable) -> None:
 
         It raises an exception if a step with the same name was already added to the workflow.
         """
+        step_config: Optional[StepConfig] = getattr(func, "__step_config", None)
+        if not step_config:
+            msg = f"Step function {func.__name__} is missing the `@step` decorator."
+            raise WorkflowValidationError(msg)
+
```

moved up as it's the most fundamental check and we can fail faster